components/authz-service/server/v2/rules_test.go

```diff
@@ -157,6 +157,149 @@ func TestCreateRule(t *testing.T) {
 	}
 }
 
+func TestUpdateRule(t *testing.T) {
+	ctx := context.Background()
+	cl, store, _, _ := setupRules(t)
+
+	apiConditions := []*api.Condition{
+		{
+			Type:     api.ProjectRuleConditionTypes_CHEF_ORGS,
+			Values:   []string{"opscode"},
+			Operator: api.ProjectRuleConditionOperators_EQUALS,
+		},
+	}
+	storageConditions := []storage.Condition{
+		{
+			Type:      storage.Node,
+			Attribute: storage.Organization,
+			Operator:  storage.MemberOf,
+			Value:     []string{"opscode"},
+		},
+	}
+
+	cases := []struct {
+		desc string
+		f    func(*testing.T)
+	}{
+		{"if the rule name is empty, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "empty-name",
+				Name:       "",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is empty, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "",
+				Name:       "empty id",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is invalid, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "no_underscores",
+				Name:       "any name",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if there are no conditions, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:        "foo",
+				Name:      "foo rule",
+				ProjectId: "bar",
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if a rule with that id does not exist, returns 'not found'", func(t *testing.T) {
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "not-found",
+				Name:       "my other foo",
+				ProjectId:  "bar",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.NotFound, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the passed rule changes the project, returns 'failed precondition'", func(t *testing.T) {
```

Comment:
you can't change the rule type, but what if you change your conditions so they no longer match the correct type?
```
NODE ATTRIBUTES
Chef Organization (string)
Chef Server (string)
Environment (string)
Chef Role (string[])
Chef Tag (string[])
Chef Policy Name (string)
Chef Policy Group (string)

EVENT ATTRIBUTES
Chef Organization (string)
Chef Server (string)
```

so if I'm editing an Event rule, I shouldn't be able to change the condition type to Chef Policy Group.

Reply:
Sure!